### PR TITLE
Adds prompt override capability (ENG-437)

### PR DESCRIPTION
**Pull request by kgrofelnik**
### Task / problem
Developer should be able to override the prompt assembly

### Changes made
- added `callbacks` to OpenCopilot (currently only `prompt_builder` member)
- added `@copilot.prompt_builder` decorator as syntactic sugar

### Testing
Try overriding prompt assembly by defining a callback function:
```python
@copilot.prompt_builder
def prompt_builder(conversation_id: UUID, user_id: str, message_str) -> str:
    docs = copilot.document_store.find(message)
    if "hello" in message.lower():
        return "Greet the user with a joke"
    return None
```
If the callback returns None, the copilot will fallback to default prompt assembly.

Callback can also be set without using a decorator in a more direct fashion:

```python
def prompt_builder(conversation_id: UUID, user_id: str, message_str) -> str:
    docs = copilot.document_store.find(message)
    if "hello" in message.lower():
        return "Greet the user with a joke"
    return None

copilot.callbacks.prompt_builder = prompt_builder
```

Simple example using available repositories:
```python
@copilot.prompt_builder
def prompt_builder(conversation_id: UUID, user_id: str, message: str) -> str:
    conversations = copilot.users_repository.get_conversations(user_id)
    history = copilot.chat_history_repository.get_messages(conversation_id)
    docs = copilot.document_store.find(message)
    print(f"This user has {len(conversations)} conversations.")
    print(f"This conversation has {len(history)} messages in history.")
    print(f"I've retrieved {len(docs)} relevant documents.")
```

### CODE REVIEW

File: opencopilot/application.py [hunk 1]
```diff
@@ -189,13 +200,18 @@ def __call__(self, *args, **kwargs):
 
     def data_loader(self, function: Callable[[], Document]):
         self.data_loaders.append(function)
+        return function
```
Comment (KasparPeterson) [inline opencopilot/application.py:203]: Why do we need to return a function in here?
If we we do need it we need to define a return type.

File: opencopilot/application.py [hunk 2]
```diff
@@ -189,13 +200,18 @@ def __call__(self, *args, **kwargs):
...
 
     def add_local_files_dir(self, files_dir: str) -> None:
         self.local_files_dirs.append(files_dir)
 
     def add_data_urls(self, urls: List[str]) -> None:
         self.data_urls.extend(urls)
 
+    def prompt_builder(self, function: PromptBuilder):
+        self.callbacks.prompt_builder = function
+        return function
```
Comment (KasparPeterson) [inline opencopilot/application.py:213]: Same here with returning function.

**Review by taivop — APPROVED** [review]
Approved as a review of interface, ignoring implementation

File: opencopilot/application.py (hunk 1 above)
Comment (kgrofelnik) [reply opencopilot/application.py:203]: If we don't then the decorated function becomes None and you can't call it explicitly anymore (not sure if we want to do that).

Without return:
```python
@copilot.prompt_builder
def builder(...):
....

print(builder)
> None

builder()
> TypeError: 'NoneType' object is not callable
```

With return:
```python
@copilot.prompt_builder
def builder(...):
....

print(builder)
> <function builder at 0x1047dc540>
```

**Pull request merged**